core/trino-main/src/main/java/io/trino/operator/IncrementalLoadFactorHashArraySizeSupplier.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import it.unimi.dsi.fastutil.HashCommon;
+
+/**
+ * Hash array size supplier that uses larger load factors for larger arrays
+ */
+public class IncrementalLoadFactorHashArraySizeSupplier
+        implements HashArraySizeSupplier
+{
+    private static final int THRESHOLD_25 = 1 << 16; // 131072
+    private static final int THRESHOLD_50 = 1 << 20; // 1048576
+
+    private final int multiplier;
+
+    public IncrementalLoadFactorHashArraySizeSupplier(int multiplier)
```

Comment:
What exactly multiplier is? How do you pick correct value at call site? Please comment out.

Reply:
It's based on TPC-DS sf1 tuning (still in progress, so subject to change)